pkg/network/tracer/gateway_lookup.go

```diff
@@ -0,0 +1,128 @@
+// +build linux_bpf
+
+package tracer
+
+import (
+	"net"
+	"unsafe"
+
+	ddconfig "github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/network"
+	"github.com/DataDog/datadog-agent/pkg/network/config"
+	"github.com/DataDog/datadog-agent/pkg/network/ebpf/probes"
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/ec2"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/ebpf"
+	"github.com/DataDog/ebpf/manager"
+)
+
+const maxRouteCacheSize = int(^uint(0) >> 1) // max int
+
+type gatewayLookup struct {
+	routeCache          network.RouteCache
+	subnetCache         map[int]network.Subnet // interface index to subnet map
+	subnetForHwAddrFunc func(net.HardwareAddr) (network.Subnet, error)
+}
+
+func gwLookupEnabled(config *config.Config) bool {
+	// only enabled on AWS currently
+	return config.EnableGatewayLookup && ddconfig.IsCloudProviderEnabled(ec2.CloudProviderName)
+}
+
+func newGatewayLookup(config *config.Config, runtimeCompilerEnabled bool, m *manager.Manager) *gatewayLookup {
+	if !gwLookupEnabled(config) {
+		return nil
+	}
+
+	var router network.Router
+	if runtimeCompilerEnabled {
+		router = newEbpfRouter(m)
+	} else {
+		router = network.NewNetlinkRouter(config.ProcRoot)
+	}
+
+	routeCacheSize := maxRouteCacheSize
+	if config.MaxTrackedConnections <= uint(maxRouteCacheSize) {
+		routeCacheSize = int(config.MaxTrackedConnections)
+	} else {
+		log.Warnf("using truncated route cache size of %d instead of %d", routeCacheSize, config.MaxTrackedConnections)
+	}
+
+	return &gatewayLookup{
+		subnetCache:         make(map[int]network.Subnet),
+		routeCache:          network.NewRouteCache(routeCacheSize, router),
+		subnetForHwAddrFunc: ec2SubnetForHardwareAddr,
+	}
+}
+
+func (g *gatewayLookup) Lookup(cs *network.ConnectionStats) {
+	r, ok := g.routeCache.Get(cs.Source, cs.Dest, cs.NetNS)
+	if !ok {
+		return
+	}
+
+	// if there is no gateway, we don't need to add subnet info
+	// for gateway resolution in the backend
+	if util.NetIPFromAddress(r.Gw).IsUnspecified() {
+		return
+	}
+
+	s, ok := g.subnetCache[r.IfIndex]
+	if !ok {
+		ifi, err := net.InterfaceByIndex(r.IfIndex)
+		if err != nil {
+			log.Errorf("error getting index for interface index %d: %s", r.IfIndex, err)
+			return
+		}
+
+		if len(ifi.HardwareAddr) == 0 {
+			// can happen for loopback
+			return
+		}
+
+		if s, err = g.subnetForHwAddrFunc(ifi.HardwareAddr); err != nil {
+			log.Errorf("error getting subnet info for interface index %d: %s", r.IfIndex, err)
```

Comment:
You'll want to make sure the feature is not enabled on agents that don't have access to AWS' metadata endpoint (non-AWS machines, or EC2 hosts that have metadata access disabled for some reason): on these environments `g.subnetForHwAddrFunc` will always return an error and the error-level log here will likely get very noisy (especially since the cache entry remains empty after an error here, so the error will be logged again on the next `Lookup` call).

You may still want to lower the log level here

Reply:
I'll use [`LogLimit`](https://github.com/DataDog/datadog-agent/blob/master/pkg/process/util/log_limit.go) here.